✨Feat(#4,#5): 모임글 이미지 업로드 기능 구현

DokyungYou opened the pull request:
## 🔎 작업 내용

### ImageController
   -  **registerGatheringImages**
      -  ```representativeImageIndex``` : 대표이미지로 설정할 인덱스
       - `int[] imageOrder`: 이미지들의 순서
       -   `List<MultipartFile> images`: 이미지파일들을 받습니다. 
        application.yml 에 최대 사이즈를 설정해놨습니다.
         ```  servlet:
            multipart:
              max-file-size: 5MB # 최대 파일 사이즈
              max-request-size: 5MB # 최대 요청 사이즈
         ```
### ImageUploadRequest
  -  컨트롤러에서 @ RequestPart 로 받아온 값들을 검증 및 서비스에 넘길 하나의 객체로 만드는 용

### ImageService
  - **registerGatheringImages**
  - **moveImageFromTempToPermanent**


### ImageOptimizeService
  - 요청받은 이미지 파일을 webp로 최적화하는 역할
    - 현재는 용량을 줄이는 역할을 생각하고 만들었습니다.)
![image](https://github.com/user-attachments/assets/2a1ed3df-c30a-4688-9795-85a49f33e927)


### S3GatheringImageService
  - 이미지파일을 받아서 s3 에 저장하는 역할
    - 파일을 바로 s3에 저장하지 않고, `서버에 file로 변환 -> webp 변환 -> s3에 저장 -> 서버에 저장됐던 file 삭제 ` 하고 있습니다. 

이미지 등록 api와 모임글 등록 api를 분리하여 구현했습니다.

### **flow**
`이미지등록 api를 먼저 요청` ->  `저장된 image url을 반환` -> `모임글 등록시 같이 넣어서 요청` -> `db에 image url 저장`

- 문제 상황:
s3 에 이미지등록은 성공했는데, 모임글 등록 과정에서 문제가 발생했을 시 해당 이미지들은 쓰이는 곳은 없는데, 저장돼있는 이미지가 됩니다. 

이 부분을 해결할 부분들을 찾아보았는데, 여러 방법이 있었습니다.  (try-catch문을 쓰거나, 배치 서버를 활용한다는 등의..)
근데 대부분 장단점이 있더군요..

일단 제 수준이랑 일정, 처음 접하는 방법을 경험해보자는 점을 고려하고나서 선택한 방법은 `s3의 수명주기 규칙 기능`을 이용하는 것이었습니다.

설정한 시간이 지나면 만료객체로 전환되게끔 했습니다. (설정시간을 더 줄이고 싶었는데 1일이 최소시간이더군요..)
![image](https://github.com/user-attachments/assets/2e352b8b-72f3-4da9-8d4f-259c33e93108)


다시 정리하자면, 
`이미지등록 api를 먼저 요청` ->  `s3의 /temp 경로에 저장` -> `url 반환 (/temp  경로의 url 아닌 수명주기가 설정돼있지 않은 영구저장 경로의 url, 아직 실제로 존재하지 않는 url )` -> `모임글 생성 api 에 넣어서 요청` -> `먼저 해당 url을 db에 넣음` -> `save하는 db 로직이 성공적으로 마침` -> `/temp 에 있던 이미지를 영구저장 경로에 복사 (이 때 db에 넣었던 url이 실체화? 되는 순간)` -> `/temp에 있던 해당 이미지는 삭제`


일단 중간에 문제가 발생할 때 생기게 되는  `사용되지 않는 이미지` 가 영구적으로 남게되는 문제는 해결된 것 같습니다..

문제는 이미지 한두 개를 요청해도 속도가 눈에 보일 정도로 느립니다. (제가 잘못 구현한거겠죠.. )

찾아보면 서버에 부담이 되는 등의 문제 때문에 서버에 파일을 보내서 요청하기보다는 s3의 presignedUrl을 활용하는 방식으로 서버리스 업로드로 구현하던데..  블로그 외의 어떤 강의 문답에서도 봤습니다..(서버에서는 presignedUrl 을 내려주는 역할만 하고 실제 파일 업로드는 프론트가.. )
하지만... 서버에서 해..했습니다.. 하하...😰

  <br/>


## 🔧  참고사항
> 중간에 설계를 몇 번이나 바꿨네요.. (이 부분에 대해서는 커밋한것도 있고 아닌것도 있습니다.)
라이브러리이랑 제공하는 클래스 선택도  중간에 자꾸 바꾸면서 별 짓을 다..

> S3GatheringImageService 에서 moveImageProcess 에서 나오는 에러를 계속 해결하고 있지 못하고 있습니다.  `Cannot invoke "Object.hashCode()" because "key" is null`
이 부분만 해결하면 되는데... 
좀 더 시도하다가 상황을 보고 계속 해결하지 못해서 일정이 지연되면 안되니까 
이때는 주석처리 해놓고 머지할 생각합니다. (다른 기능들 만들고 나서 해결할까 합니다.)


> 현재 s3에 이미지 업로드 전에 서버에 잠깐 저장하는 부분이 있는데, 임시로 프로젝트 최상단 아래에 images 라는 폴더를 만들어서 사용하고 있는데, 어디에 하는게 바람직할까요

<br/>
closed #5

DokyungYou committed:
🛠Chore(IMAGE): application.yml 수정 및 추가 (.env 값도 수정필요합니다.)

- AWS 자격 증명 속성 이름 수정: 카멜케이스 -> 스네이크케이스
- multipart 최대 파일 및 요청 크기 설정

Related to: #5

DokyungYou committed:
🛠Chore(IMAGE): build.gradle - 이미지 업로드 관련 라이브러리 의존성 추가

- io.awspring.cloud:spring-cloud-aws-starter-s3:3.1.1: s3 상호작용
- com.sksamuel.scrimage:scrimage: 이미지 최적화용

Related to: #5

DokyungYou committed:
✨Feat(IMAGE): ImageException / 관련 에러코드 / GlobalExceptionHandler - handleImageException 추가

- ImageException(ErrorCode errorCode): 직접 의도해서 던지는 예외로 사용시 사용
- ImageException(Throwable cause, ErrorCode errorCode): 이미지 로직 안에서try-catch 로 잡을 때 사용할 용도 , ErrorCode는 IMAGE_SERVER_ERROR로 사용

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): ContentImageUrlDto -> ImageUrlDto 클래스 이름 변경 (썸네일, 본문 이미지 구분하지 않고 같이 사용하게끔 변경), 관련 클래스들 수정

Related to: #5
Fixes: #4

DokyungYou committed:
✨Feat(IMAGE): ImageUploadRequest 추가

- @ RequestPart(value = "content_images", required = false) List<MultipartFile> contentImages
- @ RequestPart(value = "content_image_order", required = false) int[] contentImageOrder)
컨트롤러에서 위와 같은 형태로 받아올 것이고 파일과 순서를 매칭해서 하나의 ImageUploadRequest 로 생성

Related to: #5

DokyungYou committed:
✨Feat(IMAGE): S3GatheringImageService, ImageCompressService 추가

- S3GatheringImageService: 클라이언트로부터 받아온 MultipartFile 을 File 변환 -> webp 로 변환(썸네일은 손실압축, 본문이미지는 무손실 압축) -> s3에 업로드 -> url 반환

Related to: #5

DokyungYou committed:
✨Feat(IMAGE): ImageController, ImageService 추가

- 썸네일이미지, 본문이미지, 본문이미지 순서를 받아서 업로드

Related to: #5

DokyungYou committed:
🛠Chore(IMAGE): cloud.aws.credentials 의 key 환경변수 이름 수정

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): ImageCompressService -> ImageOptimizeService 이름 변경, convertToWebpWithLossless 주석처리

- convertToWebpWithLossless 의 경우 webp 로 바뀌는데 용량이 증가하였음
- 썸네일, 본문이미지를 구분해서 작업을 진행했었는데, 하나로 작업하게끔 계획을 변경할 예정이어서
일단은 convertToWebpWithLossless 를 사용하지 않게되어서 주석처리하였음
- 무조건 압축되는 상황이 아니어서 압축이라는 축소적 의미보다는 좀 더 넓은 의미를 담게끔 바꾸었음

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): 썸네일 이미지를 따로 받아서 처리하는 방식에서 대표이미지(썸네일) 인덱스를 받는 방식으로 수정 - ImageController, ImageUploadRequest(수정) / ImageDto(추가)

- ImageController 의 registerGatheringImages 의 파라미터 수정
- ImageUploadRequest: 개별의 파일데이터를 담던 역할에서 이미지업로드요청이라는 역할로 변경
- ImageDto: 기존의 ImageUploadRequest 의 역할을 부여

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): 썸네일 이미지를 따로 받아서 처리하는 방식에서 대표이미지(썸네일) 인덱스를 받는 방식으로 수정 - GatheringImageUrlDto -> GatheringImageUrlsDto / ImageUrlDto 주석삭제

- thumbnailUrl 필드 삭제, 대표이미지 인덱스 추가
- of 메서드 수정
- 클래스 이름을 단수형에서 복수형으로 변경
- ImageUrlDto: 주석 삭제

Related to: #5

DokyungYou committed:
♻️✨Refactor,Feat(IMAGE): 썸네일 이미지를 따로 받아서 처리하는 방식에서 대표이미지(썸네일) 인덱스를 받는 방식으로 수정 - S3GatheringImageService 수정, 메서드 추가

- 앞의 커밋들에 의한 변경: 4c58c26f / a43960d5
- processUploadContentImage: 예외처리 수정
  - 의도해서 발생시킨 ImageException 를 받을 때는 에러코드를 다시 그대로 넣어서 처리
  - 그 외의 예외는 발생한 예외와 IMAGE_SERVER_ERROR 를 넣어서 ImageException 로 처리
- moveImageProcess (모임글 등록에서 호출할 예정인 메서드) 추가
  - 기존의 temp 에 있던 이미지를 복사해서 다른 경로에 저장 후 기존의 이미지는 삭제

Related to: #5

DokyungYou committed:
✨Feat(IMAGE): 이미지 관련 에러코드 추가

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): S3GatheringImageService - registerGatheringImages 수정, moveImageFromTempToPermanent 추가

- ImageUploadRequest 로 받게끔 수정
- 썸네일 관련 로직 삭제
- 4edad389 로 인한 반환값 수정
- moveImageFromTempToPermanent: s3의 이미지 이동

Related to: #5

DokyungYou committed:
♻️Refactor(IMAGE): CreateGatheringRequest 수정

- GatheringImageUrlDto -> GatheringImageUrlsDto 변경

Related to: #5, #4

DokyungYou committed:
♻️Refactor(IMAGE): GatheringService 수정

- addContentImages: 실수로 static으로 적용했던 부분 수정
- createGatheringGeneral: 이미지 이동 로직 추가
  - 모임글이 정상적으로 등록되면, 생명주기로 관리하던 /temp 에 있던 이미지들을 영구저장하는 경로로 이동

Related to: #5, #4

DokyungYou committed:
♻️Refactor(IMAGE): 누락된 삭제 요청 추가

Related to: #5

DokyungYou committed:
Merge branch 'dev' into feature/gathering-image-upload

# Conflicts:
#	build.gradle
#	src/main/java/com/runto/global/exception/ErrorCode.java

lsy28901 reviewed (dismissed):
수고하셨습니다! 서버에 저장하는 디렉토리는 그냥 어디든 상관없다고 생각합니다.
근데 이미지 업로드를 서버를 거쳐서 하는 이유를 간단히 설명해주실 수 있나요?
성능적인 측면에서 문제가 생기는거 같은데 선택하신 이유가 궁금합니다!

DokyungYou commented:
> 수고하셨습니다! 서버에 저장하는 디렉토리는 그냥 어디든 상관없다고 생각합니다. 근데 이미지 업로드를 서버를 거쳐서 하는 이유를 간단히 설명해주실 수 있나요? 성능적인 측면에서 문제가 생기는거 같은데 선택하신 이유가 궁금합니다!

클라이언트로부터 파일을 받아와서 s3에 바로 저장하는 방식이 아닌, 
서버에 한번 file 객체로 변환 저장 하는 부분을 말씀하시는걸까요?

아니면 서버로 파일을 받아오는 것 자체를 말씀하시는걸까요?

전자의 경우라면  s3 용량 문제랑, 프론트에서 업로드시의 속도 등을 생각해서 이미지 최적화를 적용해보고 싶었는데, 변환을 하려면 먼저 file로 바꾸는 과정이 필요한 것 같더라구요.. 

후자의 의미라면.. 말을 아끼겠습니다

MateAndGit reviewed (dismissed):
수고 하셨습니다! 임시로 저장되는 파일은 리소스 하위에 두는건 어떨까요?

DokyungYou committed:
♻️Refactor(IMAGE): 이미지 임시 저장 경로 수정 (resources 폴더 하위로 이동)

- System.getProperty("user.dir") + "/src/main/resources/temp_images/"

Related to: #5

DokyungYou commented:
> 수고 하셨습니다! 임시로 저장되는 파일은 리소스 하위에 두는건 어떨까요?

옮겼습네다

lsy28901 commented:
그냥 혼자서 생각해봤는데, 왜 굳이 S3를 사용해야하는가에 대한 의문이 들었습니다. 그냥 파이어베이스 스토리지에 이미지를 저장해도 되지 않을까? 라는 생각이 들어서.. 우리는 왜 S3를 사용하는거죠?

DokyungYou commented:
> 그냥 혼자서 생각해봤는데, 왜 굳이 S3를 사용해야하는가에 대한 의문이 들었습니다. 그냥 파이어베이스 스토리지에 이미지를 저장해도 되지 않을까? 라는 생각이 들어서.. 우리는 왜 S3를 사용하는거죠?

일단 기억나는 부분은 파이어베이스 스토리지는 주로 모바일 앱에 최적화 돼있는걸로 알고 있습니다. 
기능도 상대적으로 부족하구요..? 🤔

lsy28901 commented:
S3는 대용량 파일처리에 대한 성능적인 장점이 있고, 별도의 클라우드 저장소이기 때문에 관리나 보안적 측면 등에서 장점이 있다고 들었는데, 이미지파일을 받아서 이미지를 webp로 변환하고 변환한 파일을 S3에 업로드 하는게 어떤 장점이 있을까요?

MateAndGit commented:
음 제가 생각 해봤는데 fiebase는 제가 옛날에 회사 다닐때 모바일용에 적합하다란 말이 기억이 나네요(아닐수도 있지만) 즉 그말은 대량의 데이터들 비디오 이미지등이 있겟죠? 이걸 정적 콘텐츠라고 하던데 암튼 이걸 처리하기 많이 부담스럽죠 그런데 s3는 확장이 가능해요 돈이 들지만 ㅎㅎ;; 그말은 더 효율적으로 어플리케이션을 관리할수 있구 제일 중요한 보안! aws 는 보안처리가 아주 좋습니다 그래서 지금 aws가 각광 받는 이유인거 같아요  그래서 결론적으론 빠른 개발 및 모바일 중심은 ->파베 대규모 데이터 저장이란 관리플 필요로 하는경우는 s3 + aws 인거 같습니다! 틀릴수도 있으니 참고만 하세요!

lsy28901 commented:
우리 프로젝트에서 비용상관없이 S3의 확장성과 관리적 측면을 신경쓴다면 파일변환 과정이 필요할까요?

MateAndGit commented:
비용이 상관 없다면 사실 필요가 없다고 생각해요 ㅎㅎ;; 그치만 역시 변환 과정을 거쳐서 보내면 네트워크 부담이 줄어들거 같아요!

DokyungYou commented:
> S3는 대용량 파일처리에 대한 성능적인 장점이 있고, 별도의 클라우드 저장소이기 때문에 관리나 보안적 측면 등에서 장점이 있다고 들었는데, 이미지파일을 받아서 이미지를 webp로 변환하고 변환한 파일을 S3에 업로드 하는게 어떤 장점이 있을까요?

물론 대용량 파일 처리의 장점이 있지만, 요청, 전송할때와 저장 용량에 따라 비용이 측정되기때문에, 
그 부분때문에 용량을 줄여보려 했구요.. (비용 이슈..)
브라우저에서의 이미지 로딩 속도 개선을 위해서 리사이징 하는 과정을 거치는 선택을 해보았습니다.
그리고 webp 이 효율적인 이미지 압축이 가능하다고 알고 있습니다. (웹에서의 사용을 목적으로 만들어졌다고 합니다)

MateAndGit reviewed (dismissed):
다시 승인 했습니다!

lsy28901 commented:
S3는 우리 프로젝트에 얼마나 효율적인가를 이야기 나눠보고 싶었습니다. 감사합니다!

DokyungYou commented:
이미지 최적화 등의 변환 과정이 없어도 굳이 file로 변환한 뒤에 저장하는 경우가 있긴 한 것 같긴한데.. 정확히 어떤 상황에서 쓰이고 어떤식으로 활용해야하는지는  아직도 감을 못 잡겠습니다.
![image](https://github.com/user-attachments/assets/bc3d07a5-c693-4466-8df9-58a76a6c8a15)
예전에 받아본 피드백인데, 참고해보시겠어요..?

DokyungYou committed:
🐛Fix(IMAGE): S3GatheringImageService 의 moveImageProcess 의 에러 처리 전까지의 임시방편 처리 (나중에 해결할 예정)

Related to: #5

MateAndGit reviewed (approved):
승인했습니다!!
